# HG changeset patch
# User Torsten <ttl@justmail.de>
# Date 1444986206 -7200
#      Fri Oct 16 11:03:26 2015 +0200
# Branch stable
# Node ID 802dc52d4d469d38250acb6eb88e78e105a895cf
# Parent  ffc6cdcd02c5dfb3b1c7d070e0e0c72ac4040248
improve focus detection of gui (bug #45306)

* file-editor-tab.cc (file-editor-tab.cc): set focus proxy to edit area

* file-editor.cc (enable_menu_shortcuts): explicitly set focus to editor

* octave-dock-widget.cc (octave_dock_widget): strong focus policy for all dock
  widgets

* terminal-dock-widget.cc (terminal_dock_widget): focus policy already set in
  octave_dock_widget

diff --git a/libgui/src/m-editor/file-editor-tab.cc b/libgui/src/m-editor/file-editor-tab.cc
--- a/libgui/src/m-editor/file-editor-tab.cc
+++ b/libgui/src/m-editor/file-editor-tab.cc
@@ -157,16 +157,18 @@ file_editor_tab::file_editor_tab (const 
            this, SLOT (handle_copy_available (bool)));
 
   connect (&_file_system_watcher, SIGNAL (fileChanged (const QString&)),
            this, SLOT (file_has_changed (const QString&)));
 
   QSettings *settings = resource_manager::get_settings ();
   if (settings)
     notice_settings (settings, true);
+
+  setFocusProxy (_edit_area);
 }
 
 file_editor_tab::~file_editor_tab (void)
 {
   // Destroy items attached to _edit_area.
   QsciLexer *lexer = _edit_area->lexer ();
   if (lexer)
     {
diff --git a/libgui/src/m-editor/file-editor.cc b/libgui/src/m-editor/file-editor.cc
--- a/libgui/src/m-editor/file-editor.cc
+++ b/libgui/src/m-editor/file-editor.cc
@@ -1217,16 +1217,19 @@ file_editor::enable_menu_shortcuts (bool
 {
   QHash<QMenu*, QStringList>::const_iterator i = _hash_menu_text.constBegin();
 
  while (i != _hash_menu_text.constEnd())
    {
      i.key ()->setTitle (i.value ().at (! enable));
      ++i;
    }
+
+  if (enable)
+    focus ();
 }
 
 QMenu*
 file_editor::m_add_menu (QMenuBar *p, QString name)
 {
   QMenu *menu = p->addMenu (name);
 
   QString base_name = name;  // get a copy
diff --git a/libgui/src/octave-dock-widget.cc b/libgui/src/octave-dock-widget.cc
--- a/libgui/src/octave-dock-widget.cc
+++ b/libgui/src/octave-dock-widget.cc
@@ -119,16 +119,17 @@ octave_dock_widget::octave_dock_widget (
            this, SLOT (pasteClipboard ()));
   connect (p, SIGNAL (selectAll_signal ()),
            this, SLOT (selectAll ()));
   // undo handling
   connect (p, SIGNAL (undo_signal ()), this, SLOT (do_undo ()));
 
   installEventFilter (this);
 
+  setFocusPolicy (Qt::StrongFocus);
 }
 
 octave_dock_widget::~octave_dock_widget ()
 {
   // save state of this dock-widget
   QString name = objectName ();
   QSettings *settings = resource_manager::get_settings ();
 
diff --git a/libgui/src/terminal-dock-widget.cc b/libgui/src/terminal-dock-widget.cc
--- a/libgui/src/terminal-dock-widget.cc
+++ b/libgui/src/terminal-dock-widget.cc
@@ -34,17 +34,16 @@ terminal_dock_widget::terminal_dock_widg
   terminal->setFocusPolicy (Qt::StrongFocus);
 
   setObjectName ("TerminalDockWidget");
   setWindowIcon (QIcon (":/actions/icons/logo.png"));
   set_title (tr ("Command Window"));
 
   setWidget (terminal);
   setFocusProxy (terminal);
-  setFocusPolicy (Qt::StrongFocus);
 
   connect (terminal, SIGNAL (interrupt_signal (void)),
            this, SLOT (terminal_interrupt ()));
 }
 
 bool
 terminal_dock_widget::has_focus (void) const
 {
